### ISSUE TITLE
Using Apostrophs in top-level definitions in module not working

### ISSUE THREAD

**Issue by robkuz** [fsprojects/Fable#207]
### Description

When using an apostroph in a name in a top level definition the compiled javascript the compiled module unuseable as the compiled javascript is syntactically invalid

### Repro steps

1. Define module in F#

```fsharp
module Bar
let bar x = ...
let baz y = ...

let foo' z = 1
```

2. `fable`

### Expected behavior

Export of `bar`, `baz`and `foo'`

### Actual behavior

Javascript along this lines will be generated

```javascript
exports.foo' = exports.bar = exports.baz = undefined;

//real definition ...

exports.foo' = foo_;
//# sourceMappingURL=Matrix.js.map

```
### Known workarounds

Dont use apostrophs in names

### Related information 

* Branch 0.3.22

**Comment by alfonsogarciacaro** [comment]
This should be a mistake in the RegEx I'm using to validate the public names, I'll fix it. Thanks for the report!

**Comment by alfonsogarciacaro** [comment]
Unfortunately, because of the way exports work in ES6 (they must be valid JS names) this can't be fixed. I've just added a check so an error is thrown a compile time if an invalid JS name is being used for a root member. I hope this is not much of an inconvenience.

**Comment by robkuz** [comment]
If I understand this correct?

`export["some weird' identfier"] = function () {...}`

won't work, correct?

Then what about name mangling? so that

```fsharp
let foo' = 1
let bar'' = 2
```

becomes

```javascript
var foo$apostroph =1;
var bar$apostroph$apostroph = 2;
```

and consequently then on the caller site as well.

**Comment by alfonsogarciacaro** [comment]
We can consider this (something similar is already being done for overloads), but one of Fable's goals is to make it easy to create modules that can be easily consumed from JS and this can complicate things in this regard. Please consider the problem doesn't affect only apostrophes but also any character not allowed in JS variable names.

**Comment by cosinekitty** [comment]
How about any UTF-16 code point other than [A-Za-z0-9_] gets converted to "$uxxxx", where xxxx is the code point's hexadecimal representation, similar to the "\uxxxx" encoding of JSON?  For example,
````fsharp
let ``My name$ is Earl`` = 17
````
becomes
````javascript
var My$u0020name$u0024is$0020Earl = 17;
````
It is ugly but completely predictable, and reserves the ability for other kinds of escaping via $ by using a character other than 'u' after the $.  Something like this is a necessary price for F#'s greater flexibility with variable names.

**Comment by alfonsogarciacaro** [comment]
Since there seems to be interest on this, I'm reopening the issue and changing the tag to discussion to know the opinion of others. I think the dilemma here is:

1. Allow developers to use any name in public members of root modules even if these names won't be (easily) callable from JS.
2. Force developers to use valid JS names for public members of root modules and warn them at compile time otherwise.

**Comment by robkuz** [comment]
a) yup its ugly ;-)
b) I am completely for it. The reason being that using apostroph in F# is pretty common. IF one really wants to write libraries to be consumed from pure JS he/she should refrain from using non standard identifiers. On the other hand if one writes a Fable lib (to be consumed by Fable itself) then it should be possible to just that. Just btw. I think that Purescript has a pretty similar philosophy on this issue.

**Comment by cloudRoutine** [comment]
@cosinekitty  F# is UTF-8
The following character classes are vaild for identifiers 
* [Letter, Uppercase \Lu](http://www.fileformat.info/info/unicode/category/Lu/list.htm)
* [Letter, Lowercase \Li](http://www.fileformat.info/info/unicode/category/Ll/list.htm) 
* [Letter, Titlecase \Lt](http://www.fileformat.info/info/unicode/category/Lt/list.htm)
* [Letter, Modifier \Lm](http://www.fileformat.info/info/unicode/category/Lm/list.htm) 
* [Letter, Other \Lo](http://www.fileformat.info/info/unicode/category/Lo/list.htm)
* [Number, Letter \Nl](http://www.fileformat.info/info/unicode/category/Nl/list.htm)
* [Punctuation, Connector \Pc](http://www.fileformat.info/info/unicode/category/Pc/list.htm)
* [Mark, Nonspacing \Mn](http://www.fileformat.info/info/unicode/category/Mn/list.htm)
* [Mark, Spacing Combining \Mc](http://www.fileformat.info/info/unicode/category/Mc/list.htm) 
* [Other, Format \Cf](http://www.fileformat.info/info/unicode/category/Cf/list.htm)

**Comment by alfonsogarciacaro** [comment]
This was a bit more complicated than expected but it should be fixed now, so we cannot only use apostrophes but also write function names in Japanese ;)

It was complicated because it only affected root members in the final JS file (not nested ones), including function, class and module names. I just hope I managed to cover all cases. I've followed @cosinekitty's suggestion and replaced JS non-valid ident chars with the unicode point (in hexadecimal to make it shorter) prefixed by `$`.

I'll publish shortly a new version with the fix :+1:
